### ISSUE TITLE
[5.6] Password reset token generation difference with custom hasher

### ISSUE THREAD

**Issue by tpaksu** [laravel/framework#25874]
- Laravel Version: 5.6.38
- PHP Version: 7.1.7
- Database Driver & Version: mySQL 5.7.19-winx64

#### TL;DR version

The `Illuminate\Auth\Passwords\DatabaseTokenRepository` class contains two methods which don't match at output when using a custom hasher: 

https://github.com/laravel/framework/blob/fa5ea32d59e265791918286a89cbba27305dd9c6/src/Illuminate/Auth/Passwords/DatabaseTokenRepository.php#L108-L111

and 

https://github.com/laravel/framework/blob/fa5ea32d59e265791918286a89cbba27305dd9c6/src/Illuminate/Auth/Passwords/DatabaseTokenRepository.php#L170-L173

while the custom hasher creates it's custom hash and saves in the database, the e-mail link contains the `hmac_sha256` version of the token, therefore it doesn't match at the reset password page.

    

#### Long Version

I am trying to implement pbkdf2_sha256 encryption which is used by django framework to store passwords in the database. I succeeded the login part. Now I'm trying the password reset feature. The method I used is, I created a custom hasher class by implementing `Illuminate\Contracts\Hashing\Hasher` interface, and then replaced the `Illuminate\Hashing\HashServiceProvider` with my custom service provider which replaces the `Hash` abstraction with the custom class. 

### Steps To Reproduce:

- The service provider: 

```php
<?php

namespace App\Providers;

use Illuminate\Support\ServiceProvider;

class Pbkdf2ServiceProvider extends \Illuminate\Hashing\HashServiceProvider
{
    public function register()
    {
        $this->app->singleton('hash', function () {
            return new \App\Helpers\Encryption\Pbkdf2Hasher;
        });
    }
}
```

- The encrypter class:

```php
<?php

namespace App\Helpers\Encryption;

use Illuminate\Contracts\Hashing\Hasher as HasherContract;


class Pbkdf2Hasher implements HasherContract
{
    /**
     * Gives information about the hash
     *
     * @param string $hashedValue
     * @return void
     */
    public function info($hashedValue)
    {
        if (substr_count($hashedValue, "$") === 3 && starts_with($hashedValue, "pbdkf2")) {
            list($algorithm, $iterations, $salt, $hash) = explode("$", $hashedValue);
            $algorithm = explode("_", $algorithm)[1];
            $hash = bin2hex(base64_decode($hash));
            return [
                "algorithm" => $algorithm,
                "iterations" => $iterations,
                "salt" => $salt,
                "hash" => $hash
            ];
        } else {
            return null;
        }
    }

    /**
     * Creates Django compatible hash from password using PBKDF2 hashing algorithms
     *
     * @param string $password
     * @return string $hashedPassword
     */
    public function make($value, array $options = array())
    {
        $newSalt = base64_encode(random_bytes(9));
        return 'pbkdf2_' . config("auth.pbkdf2.algorithm") . '$' . config("auth.pbkdf2.iterations") . '$' . $newSalt . '$' .
            base64_encode(hash_pbkdf2(config("auth.pbkdf2.algorithm"), $value, $newSalt, config("auth.pbkdf2.iterations")));
    }

    /**
     * Checks the value against the given hash for validation
     *
     * @param string $value
     * @param string $hashedValue
     * @param array $options
     * @return void
     */
    public function check($value, $hashedValue, array $options = array())
    {
        if (substr_count($hashedValue, "$") === 3 && starts_with($hashedValue, "pbdkf2")) {
            dd($hashedValue);
            list($algorithm, $iterations, $salt, $hash) = explode("$", $hashedValue);
            $algorithm = explode("_", $algorithm)[1];
            $hash = bin2hex(base64_decode($hash));
            return hash_pbkdf2($algorithm, $value, $salt, $iterations) === $hash;
        } else {
            return false;
        }
    }

    /**
     * Determines if the given settings does not match the settings defined in application config, thus value needs to be rehashed
     *
     * @param string $hashedValue
     * @param array $options
     * @return void
     */
    public function needsRehash($hashedValue, array $options = array())
    {
        if (substr_count($hashedValue, "$") === 3 && starts_with($hashedValue, "pbdkf2")) {
            list($algorithm, $iterations, $salt, $hash) = explode("$", $hashedValue);
            $algorithm = explode("_", $algorithm)[1];
            if (config("auth.pbdkf2.algorithm") != $algorithm) return true;
            if (config("auth.pbkdf2.iterations") != $iterations) return true;
            return false;
        } else {
            return true;
        }
    }
}
```

- Open `app/config.php` and replace this line:

```
Illuminate\Hashing\HashServiceProvider::class,
```
with
```
App\Providers\Pbkdf2ServiceProvider::class
```

- Add these lines in `config/auth.php`:

```php
    'pbkdf2' => [
        'iterations' => 36000,
        'algorithm' => 'sha256'
    ],
```

Now the hasher should work fine. Try to reset an user's password.
